src/main/kotlin/com/dprint/listeners/OptimizeJavaImports.kt

```diff
@@ -0,0 +1,83 @@
+package com.dprint.listeners
+
+import com.dprint.config.ProjectConfiguration
+import com.dprint.core.Bundle
+import com.dprint.core.LogUtils
+import com.dprint.services.editorservice.EditorServiceManager
+import com.dprint.services.editorservice.FormatResult
+import com.intellij.lang.ImportOptimizer
+import com.intellij.lang.ImportOptimizer.CollectingInfoRunnable
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.openapi.components.service
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiJavaFile
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+
+private val LOGGER = logger<OptimizeJavaImports>()
+private const val TIMEOUT = 10L
+
+/**
+ * This listener formats Java files when optimize imports called.
+ */
+class OptimizeJavaImports : ImportOptimizer {
+
+    override fun supports(file: PsiFile): Boolean {
+        if (file !is PsiJavaFile) return false
+
+        val project = file.project
+        if (!project.service<ProjectConfiguration>().state.enabled) return false
+
+        val documentManager = PsiDocumentManager.getInstance(project)
+        val document = documentManager.getDocument(file) ?: return false
+
+        val editorServiceManager = project.service<EditorServiceManager>()
+        val fileManager = FileDocumentManager.getInstance()
+        return fileManager.getFile(document)?.path?.let { editorServiceManager.canFormatCached(it) } ?: false
```

Comment:
Are we only after the path when you are using the document and file managers? I've had a bit of trouble in the past getting virtual files and have resulted to using something line the following after some investigation

```
val filePath = file.virtualFile?.path ?: file.originalFile.virtualFile?.path
```

See [this](https://github.com/dprint/dprint-intellij/blob/main/src/main/kotlin/com/dprint/formatter/DprintExternalFormatter.kt#L41-L43) for an example. Happy to keep it this way if you prefer or have good reason but figure my way might be more concise if it is sufficient.